WooCommerce/Classes/Yosemite/DefaultStoresManager.swift

```diff
@@ -70,7 +70,14 @@ class DefaultStoresManager: StoresManager {
     ///
     init(sessionManager: SessionManager) {
         _sessionManager = sessionManager
-        self.state = AuthenticatedState(sessionManager: sessionManager) ?? DeauthenticatedState()
+
+        if BuildConfiguration.shouldUseScreenshotsNetworkLayer {
+            let screenshotObjectGraph = ScreenshotObjects()
+            self.state = MockAuthenticatedState(objectGraph: screenshotObjectGraph)
+        }
+        else {
+            self.state = AuthenticatedState(sessionManager: sessionManager) ?? DeauthenticatedState()
+        }
```

Comment:
This feels like the opposite of DI here, where you depend on the state of a singleton instead of using dependency injection, so wondering if we could find a better solution 🤔 

tbh not sure if there is one that doesn't involve a (too) deep refactoring, as this would mean changing the API from Yosemite anyway 🧐 . So maybe this is fine at least for the first iteration?

I mean the idea in the end if we do DI would be to pass the `StoresManagerState` in the `init` but that would also rely on the call site to be aware of which state it is supposed to create (I imagine not _any_ `StoresManagerState` might work as expected in that context?). So not sure that would be the best solution.

Another idea would be to pass some generic factory object, which provides an `authenticatedState(from: SessionManager)` method, and let the `DefaultStoresManager.init` call that method on the factory passed as parameter to get the right one. That kinds of kicks the can down the road though.

Except maybe if we consider that this factory object I'm talking about… ends up being your `MockObjectGraph` that we rename into something more generic like `ObjectFactory` – and that would be used both for mocks and for real data, with just different instances (or conforming types) provided depending on if we're running Screenshots, UITests or real app?
This solution would be more in line with the Dependency Injection principle, but also means a deeper refactoring, so I'm curious about your thoughts: do you think it would be worth it to try this approach, or is it pushing it too far?

Reply:
>This solution would be more in line with the Dependency Injection principle, but also means a deeper refactoring, so I'm curious about your thoughts: do you think it would be worth it to try this approach, or is it pushing it too far?

I was nervous to refactor things too much with this, so I went for the simplest possible starting approach.

I agree that it'd be nicer to clean this up – I'm personally in favour of the factory object approach (because this object _determines_ the global logged-in/out state already, so we can't inject any state into it), but we can see what @shiki suggests?

Reply:
I was working on this some more today, and tried a couple of things (had a few dead-ends where I realized I was changing too much).

[One approach](https://github.com/woocommerce/woocommerce-ios/blob/e91456b21c5449be8bec80f0f8e353b1c65d3ba4/WooCommerce/Classes/Yosemite/DefaultStoresManager.swift#L106-L112) allows injecting the object graph (inside the mocked state) directly from the App Delegate's [handleLaunchArguments](https://github.com/woocommerce/woocommerce-ios/blob/e91456b21c5449be8bec80f0f8e353b1c65d3ba4/WooCommerce/Classes/AppDelegate.swift#L324-L341) method.

IMHO this is a bit cleaner, but WDYT @AliSoftware / @shiki ?

Reply:
tbh I don't really like in this approach (the one you linked in the comment above) that the `DefaultStoresManager` – which is a class belonging to the app logic/model layer) has knowledge of the Mocks (and the `MockObjectGraph` type), so that `func authenticate(withObjectGraph objectGraph: MockObjectGraph) -> StoresManager` method feels a bit off to me.

Replacing both `authenticate(credentials:)` and `authenticate(withObjectGraph:)` with a unified `func authenticate(objectGraph: ObjectGraph)` – which would pull the data from that `ObjectGraph` to init both `state` and `sessionManager.defaultCredentials` no matter what that ObjectGraph comes from, as it's not the DSM's business to know the implementation details of the OG instance it received, only what data it can pull from it – would make it more SOLID and a better separation of concern imho. Then the call site can decide to pass that method either an `ObjectGraph` that contains the real credential data (when called from the app), or an `ObjectGraph` containing mocked data (when called in the context of UITests or Screenshots). The construction of the proper `ObjectGraph` the the context could still be decided at the `handleLaunchArguments` site, completing the proper usage of the ServiceLocator pattern.

WDYT?

Reply:
Heya! Since it looks like we'll be using a different approach (https://github.com/woocommerce/woocommerce-ios/pull/3266#issuecomment-738279736). Let's probably discuss this later. 🙂 

Using that other approach seems to fix the issue with `DefaultStoresManager` having knowledge of the `MockObjectGraph`. And we could potentially even isolate everything related to Screenshots. 🤸